### PR TITLE
Errors: Hiding HTTP error messages #567

### PR DESCRIPTION
**Pull request by humboldtux**
I've started working on this, but i'm not sure i'm heading in the right direction.

I've tested it against a wordpress site, behaves as expected.

### CODE REVIEW

**Comment by alexandreteles** [comment]
I liked it. Can you provide usage cases and examples?

**Comment by humboldtux** [comment]
Thank you @alexandreteles.

You can have a look at #567 

I've reproduced the faulty behaviour by installing wordpress: when you request a non Wordpress page, the error message is appended to the html response. With this PR, nothing is appended to a non null body response; if the body is null, the message is sent; if you provide a custom error page, it is served.

**Comment by mholt** [comment]
Very nice, I think that's the right idea. Why did you copy https://github.com/mholt/caddy/blob/master/middleware/recorder.go into errors? Why not just use middleware.NewResponseRecorder?

**Comment by humboldtux** [comment]
Thanks @mholt 

size is an unexported property from middelware package.

**Comment by mholt** [comment]
Oh yeah. Well maybe we should make some changes.

Let's export middleware.responseRecorder, first of all, and then either export its fields as well *or* create some getter methods that are exported. I like the idea of Size(), Status(), Start(), etc., to get those values.

Then put this change under test and we're good to go I think!

**Comment by humboldtux** [comment]
Sounds good to me.

Where should i add tests for this PR?:

* middleware/recorder_test.go
* middleware/errors/errors_test.go
* Both

**Comment by mholt** [comment]
Since the recorder isn't really changing, let's just put the tests in errors_test.go, where the upstream response has some content already.

Even though the recorder's getter methods are just one line of code, might still be worth having a simple test for those as well; super easy :smile:

**Comment by humboldtux** [comment]
I'm back on it, sorry for the delay.

**Comment by mholt** [comment]
When you update this PR (with tests), squash the commits and I'll be able to merge it in. :+1:

**Comment by humboldtux** [comment]
I've added a simple test, not sure if it is correct/enough.

But, i have tested my PR against a wordpress site, and this is the behaviour i experience:

With the following Caddyfile

```
0.0.0.0:8080
root wordpress
rewrite {
    to {path} {path}/ /index.php?url={uri}
}
errors {
   404 404.html
}
fastcgi / /var/run/php5-fpm.sock php
```

Without my PR, Caddy appends the content of 404.html to the body; With my PR, Caddy appends nothing to the body of th 404 page sent by wordpress.

But if i omit the errors block:

```
0.0.0.0:8080
root wordpress
rewrite {
    to {path} {path}/ /index.php?url={uri}
}
fastcgi / /var/run/php5-fpm.sock php
```

With or without my PR, Caddy always appends its default error message.

Shouldnt this PR always prevent Caddy from appending anything even if the erros block is omitted to solve #567

**Comment by mholt** [comment]
Oh, I was just wondering about that. Thanks for a clear reproduction of the issue.

It would be nice to fix it in both places. I wonder if we can reliably do that without having to buffer every single response. Maybe by looking for a Content-Length header or something, since that (should) only be set if content has been written to the response already. Does that work?

**Comment by humboldtux** [comment]
Sorry being that dumb: where should the Content-Length check should take place? in middleware/errors/errors.go?

**Comment by mholt** [comment]
Yeah, and in the server.go file as well, where the 'default' error handler is. Before writing an error to the response, we should check if Content-Length is not empty, and if it's not empty, that means a response has already been written.

**Comment by humboldtux** [comment]
Ok, i think i understand where to make this change https://github.com/mholt/caddy/blob/master/server/server.go#L323

But i'm not sure it's possible to get Content-Length, sorry but don't forget it's a `good-for-newbie` issue :smile:

**Comment by mholt** [comment]
Sorry! I didn't think it would get this complicated. But getting the header is very easy, just call `w.Header().Get("Content-Length")`. See if that works, and let me know!

**Comment by humboldtux** [comment]
Ok, with this PR everything works as expected.

I couldn't use `w.Header().Get("Content-Length")` in middleware/errors/errors.go because tests were failing: is it correct?

**Comment by humboldtux** [comment]
Hold on @mholt , something is broken

**Comment by humboldtux** [comment]
every use cases has been tested and works now:
1) errors middleware + no middleware
2) errors middlware + middleware (fastcgi)
3) without errors middleware + middleware (fastcgi)
4) no middleware at all (launching caddy without a Caddyfile)

1 and 4 weren't working in the previous PR, i had to change:

`w.Header().Get("Content-Length") == "0"`

to
 
`w.Header().Get("Content-Length") == ""`

But tests were passing: what did i miss? What test are missing to test this edge case?

**Comment by humboldtux** [comment]
I'm sorry Matt, i'm tired, wait a little to merge in.

**Comment by abiosoft** [comment]
@humboldtux 
No need to be sorry :) This is not a race and there is no deadline. Take it easy.

Thanks for working on this.

**Comment by humboldtux** [comment]
@abiosoft thank you for your kind words

**Comment by humboldtux** [comment]
The changes to server/server.go are not enough: using a middleware like fastcgi without using errors middleware(use case 3) appends Caddy default error to Body.

**Comment by mholt** [comment]
@humboldtux Really? Even though content-length is empty? That means that Content-Length is not getting set... Can you figure out why and see if there's a way to make that work?

File: middleware/errors/errors.go [hunk 1]
```diff
@@ -36,14 +37,17 @@ func (h ErrorHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, er
...
 		}
 		h.Log.Println(errMsg)
 	}
 
 	if status >= 400 {
-		h.errorPage(w, r, status)
+		// Check there is no response body already sent
+		if responseRecorder.Size() == 0 {
+			h.errorPage(responseRecorder, r, status)
+		}
```
Comment (mholt) [inline middleware/errors/errors.go]: Now that your PR has been merged, we can just check the content-length header like you do below; no need to make a ResponseRecorder. :+1:

**Comment by mholt** [comment]
I actually just went ahead and built on your changes to finish fixing this since I know our time zones are so different :clock230: :wink: Thank you for seeing this through!

When you get a chance, could you pull down master and make sure my commit fixes it with a real PHP backend? (I just wrote a test for it.)

**Comment by humboldtux** [comment]
Ok for me: I tested master and Caddy error messages are not appended anymore if not needed.

Fortunately, this was just a `good-for-newbies` issue :smile:  Thank you for your patience.

**Comment by mholt** [comment]
Awesome :+1: Really appreciate your help!

**Comment by slightfoot** [comment]
This has regressed. Issue still exists since moving from older middleware system.
https://github.com/mholt/caddy/blob/master/caddyhttp/errors/errors.go#L53

Previous mholt patch:
https://github.com/mholt/caddy/commit/c37ad7f677b9122544ab2d2384ef8ef2b76bd970

Previous PR patch:
https://github.com/mholt/caddy/pull/614/files#diff-2133a138b5410dfb36af9638df4e783fL45

I noticed this when creating my own middleware sending back JSON body and having the default error handler injecting its error text. I'd also like to go on to say in simple error reporting from middleware you might not want to set a Content-Length header. So recording the number of bytes written to the http response would probably be better.
